### ISSUE TITLE
[tf1.15] Low GPU usage with multi gpus and multi losses  

### ISSUE THREAD

**Issue by Yablon** [tensorflow/tensorflow#42881]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: None
- TensorFlow installed from (source or binary): build from source according to official documents. branch is r1.15
- TensorFlow version (use command below): 1.15
- Python version: 2.7
- Bazel version (if compiling from source): 0.24.1
- GCC/Compiler version (if compiling from source): 5.4.0
- CUDA/cuDNN version: 10.1 / 7.6.2
- GPU model and memory: v100, 16G

**Describe the current behavior**
I trained the model belowing with 2 v100 GPUs, each of which has 16GB memory, and get
Low GPU usage, average is 20 percent.

I tried batch size 384 on single GPU, the usage does not go any higher. I think it means that the low usage of GPU has nothing to do with batch size.

And I tried on colab, [results](https://colab.research.google.com/drive/13N7ukulsc06WWf8B079jQIcEpQ7zUVYt?usp=sharing). Although I can't get the GPU usage while training, but the time by each step is the same with my local experiments.

**Describe the expected behavior**
High GPU usage and faster training.

**Standalone code to reproduce the issue**
```
import numpy as np
import tensorflow as tf
from tensorflow.keras import backend as K
from tensorflow.keras.callbacks import TensorBoard
from tensorflow.keras.layers import GRU, CuDNNGRU, Dense, Embedding, Reshape
from tensorflow.keras.optimizers import Adam

batch_size = 256
time_steps = 750

# datasets with almost no time delay
np_input = np.random.random((batch_size, time_steps, 512))
np_output = np.random.randint(0, 32, size=(batch_size, time_steps, 8))
def generate():
  while True:
    yield np_input, np_output
output_shapes = ((batch_size, time_steps, 512), (batch_size, time_steps, 8))


# split the output to 8 parts, calculate loss with 8 labels
def loss(y_true, y_pred):
  y_true = tf.split(y_true, num_or_size_splits=8, axis=-1)
  y_pred = tf.split(y_pred, num_or_size_splits=8, axis=-1)
  loss_func = lambda true, pred: tf.keras.losses.sparse_categorical_crossentropy(
      true, pred, from_logits=True)
  return tf.reduce_mean(
      tf.add_n([loss_func(y_true[i], y_pred[i]) for i in range(8)]))


# two v100 gpus, each of which has 16GB of memory
strategy = tf.distribute.MirroredStrategy()
with strategy.scope():
  # get dataset
  dataset = tf.data.Dataset.from_generator(generate,
                                           output_types=(tf.float32, tf.int32),
                                           output_shapes=output_shapes)
  dataset = dataset.prefetch(2)

  # create model
  inputs = tf.keras.Input(shape=(None, 512))
  rnn_outputs, rnn_state = CuDNNGRU(384,
                                    return_sequences=True,
                                    return_state=True)(inputs)
  h_1, h_2 = tf.split(rnn_outputs, num_or_size_splits=2, axis=-1)
  logits_1 = Dense(128)(Dense(256, activation='relu')(h_1))
  logits_2 = Dense(128)(Dense(256, activation='relu')(h_2))
  outputs = K.concatenate([logits_1, logits_2], axis=-1)
  model = tf.keras.Model(inputs, outputs)

  tensorboard = TensorBoard(log_dir='/workspace/exp/tf_log',
                            update_freq=500,
                            profile_batch=0)
  optimizer = Adam(lr=1e-4, clipnorm=1.)
  model.compile(optimizer=optimizer, loss=loss)

# train
model.fit(dataset, epochs=4, steps_per_epoch=1000, callbacks=[tensorboard])

```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

I split the outputs of CudnnGRU to 2 parts, and split the final logits to 8 parts to calculate loss. Is there any possible problems in my code ? If any, how can I improve the gpu usage ?

**Comment by ravikyram** [comment]
I have tried in colab with TF version 1.15.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/73a2948e34416502ff587892b945161b/untitled294.ipynb). Thanks!

**Comment by Yablon** [comment]
@ravikyram hi, I tried to remove the tf.distribute.MirroredStrategy and trained on a single GPU, and that problem did not disappear.

**Comment by Yablon** [comment]
@ravikyram I tried to simplify the codes, the following code leads to low usage of GPU on a 16G v100 too.

```
import numpy as np
import tensorflow as tf
from tensorflow.keras import backend as K
from tensorflow.keras.callbacks import TensorBoard
from tensorflow.keras.layers import GRU, CuDNNGRU, Dense, Embedding, Reshape
from tensorflow.keras.optimizers import Adam

batch_size = 384
time_steps = 750

# datasets with almost no time delay
np_input = np.random.random((batch_size, time_steps, 512))
np_output = np.random.randint(0, 32, size=(batch_size, time_steps, 1))


def generate():
  while True:
    yield np_input, np_output


output_shapes = ((batch_size, time_steps, 512), (batch_size, time_steps, 1))


# split the output to 8 parts, calculate loss with 8 labels
def loss(y_true, y_pred):
  loss_func = lambda true, pred: tf.keras.losses.sparse_categorical_crossentropy(
      true, pred, from_logits=True)
  return loss_func(y_true, y_pred)


# two v100 gpus, each of which has 16GB of memory
# get dataset
dataset = tf.data.Dataset.from_generator(generate,
                                         output_types=(tf.float32, tf.int32),
                                         output_shapes=output_shapes)
dataset = dataset.prefetch(3)

# create model
inputs = tf.keras.Input(shape=(None, 512))
rnn_outputs, rnn_state = CuDNNGRU(384, return_sequences=True,
                                  return_state=True)(inputs)
logits_1 = Dense(128)(Dense(256, activation='relu')(rnn_outputs))
model = tf.keras.Model(inputs, logits_1)

optimizer = Adam(lr=1e-4, clipnorm=1.)
model.compile(optimizer=optimizer, loss=loss)

# train
model.fit(dataset, epochs=4, steps_per_epoch=1000)
```

**Comment by Yablon** [comment]
Here is the [colab link](https://colab.research.google.com/drive/1j8CqU_7LoRqt8iIBlw2JRCpegN9OtBU8?usp=sharing) for the simplified code.

**Comment by nikitamaia** [comment]
Hi @Yablon, we are currently working on a performance debugging guide that will be added to the TF docs and hopefully help for issues like this. Until that's out I can try to help you here.

Can you provide more detail on how you measured your GPU utilization?

When debugging performance issues with a distribution strategy, it's useful to start by making sure you have optimal performance in the single GPU case first as you've noted in your comment above. 

The first thing to do in these cases it to use the Tensorboard Profiler. With the trace viewer, we can get an idea of what could be causing the problem. Here is an [overview](https://www.tensorflow.org/guide/profiler) of how to use the Profiler, and here is a [simple example](https://www.tensorflow.org/tensorboard/tensorboard_profiling_keras) for using the profiler with a single GPU.  If you can provide some screenshots based on the performance (especially the trace viewer) that will help to debug.

I am also curious to know what you see when you replace the CuDNNGRU unit with another layer. Additionally, it also seems like your code should run fine with TF 2.x (with the exception of the CuDNNGRU layer) so I would be interested to know what the results look like in that case as well.
